# Binary Search
Binary search, also known as logarithmic search, is a searching algorithm that finds a target value in sorted array by repeatedly dividing the search innterval by half. It's very important to note that the Binary Search algorithm only works provided the list/array is sorted.  
This searching algorithm is quite superior to linear search in terms of time complexity and space complexity.<br>

The logic behind this searching algorithm is very simple;  
Given a sorted array, intervally begin with the mid element,<br>
If the element to be found is equal to the mid element, then return the index of the mid element,<br>
Or if the element to be found is less than the mid element of the interval, narrow the interval to the lower half.  
Else, narrow it to the upper half.  
This process is done iteratively or recursively until the element is found or the interval is empty.<br>  

I'll be implementing the iterative method of binary search, as well as the recursive method of binary search.  
Also, I will implement a linear search just to compare the time taken for executing both search algorithms.  
Let's get to it....

In [1]:
# Import time_it for time comprison.
from utils import timely


@timely
def linear_search(array, search_key):
    for index, element in enumerate(array):
        if element == search_key:
            return index
    return -1

@timely
def binary_searchI(array, search_key):
    # Minimum index.
    mini = 0
    # Maximum index.
    maxi = len(array) - 1
    
    while mini <= maxi:
        # Middle index.
        midi = (mini+maxi)//2
        mid_num = array[midi]
        
        if mid_num == search_key:
            return midi
        
        if mid_num < search_key:
            mini = midi + 1
        else:
            maxi = midi - 1
            
    return -1


def binary_searchR(array, search_key, mini, maxi):
    if maxi < mini:
        return -1
    
    midi = (mini+maxi)//2
    if midi >= len(array) or midi < 0:
        return -1
    
    mid_num = array[midi]
        
    if mid_num == search_key:
        return midi

    if mid_num > search_key:
        maxi = midi - 1
    else:
        mini = midi + 1
    
    return binary_searchR(array, search_key, mini, maxi)

The linear search is just a simple for loop iterating over each element in the proposed array, and once it finds an element equal to the search key, it returns the index of that element.<br>  
From the above, we can see that there are two binary search function; `binary_searchI` and `binary_searchR`.<br>
The I and R stands for Iterative and Recursive respectively.  
Both functions follow the logic I explained earlier, in which the middle number is found, then depending on whether the search key is greater or less than the middle number, an appropriate half of the array is discarded.  
If the search key is greater than the middle number, the left/lower half is discarded, otherwise, the right/upper half is discarded.  
These two functions are essentially the same, except one performs its operation iteratively, while the other performs its operations recursively.<br>  
__NB:__ In binary search algorithm, the search key is always found at the `mid_num` of an interval. However, it most likely won't be the first or second interval.<br>  
I decorated `linear_search` and `binary_searchI` with `timely`, a decorator that returns how much time each operation takes to execute.  
Also, I did not decorate `binary_searchR` because it is a recursive function. Each recursion in a recursive operation will trigger the decorate and make the output messy.<br>  
Now, let's see compare the linear search and binary search algorithms....

In [2]:
if __name__ == "__main__":
    print("This is a demo of comparison between linear search and binary search, from ifunanyaScript.")
    print("_" * 90)
    array = [*range(1, 200000000)]
    search_key = 50056
    # Linear searching
    index = linear_search(array, search_key)
    print(f"Search key found at index {index} using linear search.")
    print("_" * 51)
    print("\n")
    index = binary_searchI(array, search_key)
    print(f"Search key found at index {index} using binary search.")

This is a demo of comparison between linear search and binary search, from ifunanyaScript.
__________________________________________________________________________________________
linear_search took 2.38395 milli-seconds to execute.
Search key found at index 50055 using linear search.
___________________________________________________


binary_searchI took 0.05865 milli-seconds to execute.
Search key found at index 50055 using binary search.


Viola!!!<br>
It's obvious which algorithm is more efficient.<br>
Let's test the recursive function...

In [4]:
binary_searchR(array, search_key, mini=0, maxi=(len(array)-1))

50055

Now, there could be a scenario where the search key occurs more than once in array. For example, 50056 could appear 4 times in the array, and each will occupy indices; 50055, 50054, 50056, 50057.  
Using the binary search functions above, these other ocuurances will be missed.  
Hence, I'll be writing a function that addresses this quandry and returns the indices of all occurances.  
Let's get to it...

In [4]:
binary_searchI(array, search_key)

binary_searchI took 0 mili-seconds


909099988

In [19]:
(4+6)//2

5